feat(typeahead): allow custom window and no result templates

fcsa-teamhammer opened the pull request:
Fixes #1619
Closes #1619

dmytroyarmak committed:
fix(typeahead): resubscribe for value changes on blur, esc, enter

Fixes #723
Closes #1244

jiayihu committed:
feat(accordion): support nodes preservation

Closes #1370

PerfectPixel committed:
fix(datepicker): propagate model and validation state on 'input' event

Fixes #1225
Closes #1605

ktriek committed:
feat(tabset): add nav justification and orientation

Closes #1472
Closes #1239

Commit:
feat(typeahead): allow custom window and no result templates

Commit:
feat(typeahead): allow custom window and no result templates

Commit:
fix(typeahead): Fixing noResultsTemplate window content tests.

Commit:
fix: close typeahead window no results and template

Tim-SE commented:
We are also waiting for this pull request to get implemented as we need the custom window ability to utilize headers in our drop-down window.  We have tested this in our application and it meets our needs.

pkozlowski-opensource committed:
chore: prepare branch for Bootstrap 4 beta.1 update

pkozlowski-opensource committed:
refactor: CSS class rename card-block -> card-body

pkozlowski-opensource committed:
refactor: CSS class rename btn-secondary -> btn-outline-secondary

pkozlowski-opensource committed:
demo: fix API / Examples tab colors for Bootstrap 4 beta

pkozlowski-opensource committed:
refactor(carousel): remove unnecessary left / right CSS classes

pkozlowski-opensource committed:
demo: update responsive utility classes

pkozlowski-opensource committed:
chore: update Bootstrap CSS to the latest published version

pkozlowski-opensource committed:
refactor(popover): update CSS class names and arrow positioning

pkozlowski-opensource committed:
refactor(tooltip): rename Bootstrap CSS classes, add arrow element

pkozlowski-opensource committed:
fix(popover): account for margins when positioning popovers

rahul-intel commented:
Please resolve the conflict ASAP as we want to use custom window ability to utilize headers in our typeahed dropdown.
Thanks

Commit:
fix(typeahead): resolving merge conflicts

Commit:
fix(typeahead): update comment and formatting

Commit:
fix(typeahead): update comment and formatting

rmeans commented:
I have merged in the latest code from master and resolved all of the conflicts.

jnizet committed:
demo(datepicker): remove radios and just show French datepicker

Fixes #1712
Closes #1723

pkozlowski-opensource committed:
demo(buttons): properly generate paths for plunkers

Fixes #1714

Closes #1724

pkozlowski-opensource committed:
fix(datepicker): retain manually entered values even if invalid

Fixes #1710

Closes #1725

pkozlowski-opensource committed:
demo(datepicker): use buttons for calendar icon

`<div>`s are not focusable with keyboard so let's use
`<buttons>`s instead.

Closes #1726

pkozlowski-opensource committed:
chore: release 1.0.0-alpha.30

maxokorokov committed:
refactor(datepicker): replace secondary/faded classes with light

maxokorokov committed:
demo: change navbar brand/button order, so brand would be on the right

pkozlowski-opensource committed:
feat(typeahead): add support for the 'container' input

Closes #1105

Closes #1731

pkozlowski-opensource committed:
feat(buttons): allow vertical radio buttons

Closes #1231
Closes #1238

BREAKING CHANGE:

The `btn-group` CSS class needs to added explicitly for radio buttons.

Before:

```
<div ngbRadioGroup ...>
  ...
</div>
```

After:

```
<div class="btn-group" ngbRadioGroup ...>
  ...
</div>
```

Closes #1733

markmssd commented:
Would this PR give the ability to have grouped result items, each with its own header?

GJ @fcsa-teamhammer 🥇

jnizet committed:
chore: use bootstrap 4.0.0-beta

Closes #1739

jnizet committed:
demo: display all alert types, and mention them all in doc

Closes #1740

jnizet committed:
demo: add a hr in two demos for better vertical spacing

Closes #1742

pkozlowski-opensource committed:
feat(dropdown): distinguish inside and outside clicks for autoClose

Closes #1022

BREAKING CHANGE:

Dropdown menu now requires usage of the new `ngbDropdownMenu` directive.

Before:

```html
<div ngbDropdown>
  <button ngbDropdownToggle>Toggle dropdown</button>
  <div class="dropdown-menu">
    <a class="dropdown-item">Action</a>
  </div>
</div>
```

After (notice **ngbDropdownMenu**):

```html
<div ngbDropdown>
  <button ngbDropdownToggle>Toggle dropdown</button>
  <div ngbDropdownMenu>
    <a class="dropdown-item">Action</a>
  </div>
</div>
```

jnizet committed:
demo: fix validation feedback

The success and danger messages are now displayed properly, but the timepicker input fields still don't have the appropriate border.

Closes #1744

pkozlowski-opensource committed:
fix(dropdown): update for Bootstrap4.beta

Closes #1745

pkozlowski-opensource committed:
chore: release 1.0.0-alpha.31

pkozlowski-opensource committed:
chore: switch to 4.0.0-beta

BREAKING CHANGE:

ng-bootstrap now requires Bootstrap CSS 4.0.0-beta

Closes #1679

pkozlowski-opensource committed:
chore: release 1.0.0-beta.1

rmeans commented:
@markmssd no unfortunately it would not.

roger2hk committed:
docs(README): update Bootstrap 4 link

Closes #1750

pkozlowski-opensource committed:
fix(dropdown): restore support for dropups

Fixes #1747
Part of #1171
Part of #1012

BREAKING CHANGE:

The `up` input is no longer supported by you can use more flexible
`placement` setting now.

Before:

```html
<div ngbDropdown [up]="true">
```

After:

```html
<div ngbDropdown placement="top-right">
```

Closes #1752

pkozlowski-opensource committed:
fix(dropdown): correct TS typings

ref: https://stackoverflow.com/q/45716754/1418796

Closes #1754

pkozlowski-opensource committed:
docs: add warning about jQuery / Bootstrap JS

Closes #1749

Closes #1757

pkozlowski-opensource committed:
fix(accordion): allign with Bootstrap 4.beta markup

Closes #1050
Closes #1332

pkozlowski-opensource committed:
chore: release 1.0.0-beta.2

Commit:
chore: fix npm test command

Closes #1770

mschoudry committed:
feat: add support for "auto" placement and multi-placement

Closes #992
Closes #1782

mschoudry committed:
feat: support more positioning options for tooltips and popovers

Closes #857
Closes #1772

Commit:
demo: make examples tab default and fix mobile friendly display

Closes #1781

Commit:
demo: api page styling improvement

Closes #1790

Commit:
demo: memorize selected tab (examples / api)

Closes #1796

pkozlowski-opensource committed:
fix: restore compatibility with 5.x

Closes #1791

Closes #1792

pkozlowski-opensource committed:
chore: release 1.0.0-beta.3

kpwbo committed:
feat(modal): add 'beforeDismiss' option

Closes #1598

Closes #1799

Commit:
feat(demo): examples page UX improvements

Closes #1801

maxokorokov committed:
demo: add colored badges for items in documentation

Closes #1806

pkozlowski-opensource committed:
demo(dropdown): show usage of split and group buttons

Closes #1785
Closes #1809

kpwbo committed:
demo: fix github buttons deprecation warnings

Closes #1812

mschoudry committed:
feat(dropdown): auto and array support in placement

Closes #1171
Closes #1814

mschoudry committed:
feat(datepicker): add support for placement "auto"

Closes #1257
Closes #1820

mschoudry committed:
fix(positioning): check for secondary available placements

Closes #1819

mschoudry committed:
feat(typeahead): added support for placement

Closes #1767
Closes #1821

NathanHazout committed:
refactor(datepicker): remove confusing comment

Closes #1822
Closes #1827

pkozlowski-opensource committed:
chore: release 1.0.0-beta.4

pkozlowski-opensource committed:
fix(modal): remove unnecessary Injectable annotation

Fixes #1832

Closes #1833

sibiraj-s committed:
docs: fix typo

Closes #1835

pkozlowski-opensource committed:
demo: avoid duplicate ids in generated HTML

Closes #1838

maxokorokov committed:
refactor(accordion): remove redundant internal state

Closes #1843

mschoudry committed:
fix(dropdown): add "dropdown" class for placements other than "top"

Fixes #1847
Closes #1852

pkozlowski-opensource committed:
chore: release 1.0.0-beta.5

mschoudry committed:
fix(popover): added margin to arrow placement

Fixes #1845
Closes #1855

mschoudry committed:
feat(datepicker): add support for the container option

Closes #1380
Closes #1857

chenyuzhcy committed:
fix(datepicker): popup disabled state

kiLLeen commented on src/typeahead/typeahead-window.ts:
```diff
@@ -73,12 +133,26 @@ export class NgbTypeaheadWindow implements OnInit {
+  _getWindowContext() { return this._context; }
+
+
+  _getNoResultsContext() { return {term: this.term}; }
```
I'm unsure of why this method (and the one above it) is prefixed with an underscore. Is it because the method is a getter for a private member variable?

fcsa-teamhammer replied:
I believe this was the pattern already established in the file to indicate the private function which is used by the template to get the context for the two different ng-template sections. The contexts are not intended to be exposed outside of the control.

kiLLeen commented:
Where are we on this PR? Just looks like some small conflicts here. Great job here @fcsa-teamhammer

maxokorokov committed:
refactor(datepicker): simplify disabled state handling

Fixes #1829

fcsa-teamhammer commented:
We are wondering the same thing. Looking at the list of pull requests it seems like there is not much appetite to add in community generated functionality. It would be nice to have the project owners @pkozlowski-opensource establish a vision for contributions to help set expectations. We are eager to update our pull request and get it in, but I don't want to continue to make fixes with the hope that one day it will be pulled in.

Commit:
feat(typeahead) updating to use template outlet context

rmeans committed:
Merge branch 'master' into typeahead-custom-templates

Commit:
fix(typeahead): fixing bad merge

paulbargaoanu commented:
Hi, everyone! Does anyone know if any progress has been made on this? I can see that #1625 failed, however this PR looks like it's ready for merging.  It would really be great if we could add some static lines (e.g. header, footer) to the typeahead pop under. Alternatively, if this feature cannot be supported by the Bootstrap Team, has anyone found a workaround for this? I would deeply appreciate any tips. Thanks a lot!

rmeans committed:
Merge branch 'master' into typeahead-custom-templates

rcuddy commented:
Adding in my vote for this as well; result formatter isn't enough, we need control over the window component as well, which this PR seems to provide. Why has this not been merged yet? :(

anytimecoder commented:
Any reason why this PR wasn't merged at the time it didn't have any conflicts? This would be a very useful feature to have.

fcsa-teamhammer commented:
@anytimecoder we were wondering the same thing. I gave up trying to keep the branch up to date because it didn't appear the PR was ever going to be brought in.

gentoo90 commented:
@ng-bootstrap, does this PR need any more fixes to be merged?

MixonZ commented:
Any update?

panic175 commented:
Is there any update regarding this? I would like to have a window which supports virutal scrolling which is would only be possible if I'm able to replace the results window completely.
